Generate serialization of CFArrayRef

achristensen07 opened the pull request:
#### 817b93374cbff38a0b7693d8b316219157b68478
<pre>
Generate serialization of CFArrayRef
<a href="https://bugs.webkit.org/show_bug.cgi?id=268700">https://bugs.webkit.org/show_bug.cgi?id=268700</a>

Reviewed by Brady Eidson.

To retain existing behavior, we need a type distinct from CoreIPCArray which wraps NSArray.
The day may come when all our array serialization specifies what is inside the array, but
until then we need CoreIPCCFArray.

Covered by tests from 274048@main

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
(IPC::ArgumentCoder&lt;CoreIPCRetainPtr&lt;T&gt;&gt;::encode):
(IPC::CoreIPCRetainPtr::CoreIPCRetainPtr): Deleted.
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::encodeObjectDirectly&lt;CFTypeRef&gt;):
* Source/WebKit/Shared/Cocoa/CoreIPCCFType.h:
* Source/WebKit/Shared/Cocoa/CoreIPCCFType.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCRetainPtr.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCCFType.h.
(IPC::CoreIPCRetainPtr::CoreIPCRetainPtr):
* Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
(IPC::typeFromCFTypeRef):
(IPC::ArgumentCoder&lt;CFArrayRef&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;RetainPtr&lt;CFArrayRef&gt;&gt;::decode): Deleted.
* Source/WebKit/Shared/cf/ArgumentCodersCF.h:
* Source/WebKit/Shared/cf/CFTypes.serialization.in:
* Source/WebKit/Shared/cf/CoreIPCCFArray.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCCFType.h.
(WebKit::CoreIPCCFArray::CoreIPCCFArray):
(WebKit::CoreIPCCFArray::array const):
* Source/WebKit/Shared/cf/CoreIPCCFArray.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCCFType.h.
(WebKit::CoreIPCCFArray::CoreIPCCFArray):
(WebKit::CoreIPCCFArray::createCFArray const):
* Source/WebKit/Shared/cf/CoreIPCCFArray.serialization.in: Copied from Source/WebKit/Shared/Cocoa/CoreIPCCFType.serialization.in.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/274062@main">https://commits.webkit.org/274062@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43e336ab2d8b1b94fc29d3a8a5a2013745dd58d0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37750 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16645 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39992 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40291 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33588 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19292 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13863 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31945 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38317 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14001 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33043 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12240 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12170 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41550 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34146 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34168 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38061 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12769 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10326 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36236 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14183 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13151 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4895 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/13494 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/Shared/cf/CoreIPCCFArray.mm:
```diff
@@ -0,0 +1,57 @@
+
+} // namespace WebKit
+
+#endif // PLATFORM(COCOA)
```
```suggestion
#endif // USE(CF)
```

aprotyas reviewed (commented):
Minor comment

beidson reviewed (approved):
Some of the xcodeproj changes feel unrelated.

achristensen07 committed:
Generate serialization of CFArrayRef
https://bugs.webkit.org/show_bug.cgi?id=268700

Reviewed by Brady Eidson.

To retain existing behavior, we need a type distinct from CoreIPCArray which wraps NSArray.
The day may come when all our array serialization specifies what is inside the array, but
until then we need CoreIPCCFArray.

Covered by tests from 274048@main

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
(IPC::ArgumentCoder<CoreIPCRetainPtr<T>>::encode):
(IPC::CoreIPCRetainPtr::CoreIPCRetainPtr): Deleted.
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::encodeObjectDirectly<CFTypeRef>):
* Source/WebKit/Shared/Cocoa/CoreIPCCFType.h:
* Source/WebKit/Shared/Cocoa/CoreIPCCFType.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCRetainPtr.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCCFType.h.
(IPC::CoreIPCRetainPtr::CoreIPCRetainPtr):
* Source/WebKit/Shared/cf/ArgumentCodersCF.cpp:
(IPC::typeFromCFTypeRef):
(IPC::ArgumentCoder<CFArrayRef>::encode): Deleted.
(IPC::ArgumentCoder<RetainPtr<CFArrayRef>>::decode): Deleted.
* Source/WebKit/Shared/cf/ArgumentCodersCF.h:
* Source/WebKit/Shared/cf/CFTypes.serialization.in:
* Source/WebKit/Shared/cf/CoreIPCCFArray.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCCFType.h.
(WebKit::CoreIPCCFArray::CoreIPCCFArray):
(WebKit::CoreIPCCFArray::array const):
* Source/WebKit/Shared/cf/CoreIPCCFArray.mm: Copied from Source/WebKit/Shared/Cocoa/CoreIPCCFType.h.
(WebKit::CoreIPCCFArray::CoreIPCCFArray):
(WebKit::CoreIPCCFArray::createCFArray const):
* Source/WebKit/Shared/cf/CoreIPCCFArray.serialization.in: Copied from Source/WebKit/Shared/Cocoa/CoreIPCCFType.serialization.in.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/274062@main

mscottapple commented on Source/WebKit/Shared/cf/CoreIPCCFArray.mm:
```diff
@@ -0,0 +1,57 @@
+    for (CFIndex i = 0; i < count; i++) {
+        CFTypeRef element = CFArrayGetValueAtIndex(array, i);
+        if (IPC::typeFromCFTypeRef(element) == IPC::CFType::Unknown)
+            continue;
```
Just silently dropping this seems unwise. Definitely could make debugging tricky.
